url1f1de5fc71_cooijmanstim_hobo_tgz-pJ8-hobo_TestingEnviromentJ8/src/hobo/Belief.java

```diff
@@ -0,0 +1,636 @@
+		double p = 1;
+
+		CardBag unknown = cards_of_unknown_location.clone();
+
+		CardBag unknown_deck_cards = s.deck.clone();
+		assert(unknown_deck_cards.containsAll(known_deck_cards));
+		unknown_deck_cards.removeAll(known_deck_cards);
+
+		p *= unknown.probabilityOfSample(unknown_deck_cards);
+		unknown.removeAll(unknown_deck_cards);
+
+		for (int i = 0; i < players.length; i++) {
+			if (i == player)
+				continue;
+
+			CardBag unknown_hand_cards = s.playerState(i).hand.clone();
+			assert(unknown_hand_cards.containsAll(players[i].known_cards));
+			unknown_hand_cards.removeAll(players[i].known_cards);
+
+			p *= unknown.probabilityOfSample(unknown_hand_cards);
+			unknown.removeAll(unknown_hand_cards);
+		}
+		
+		return p;
+	}
+	
+	// modifies s
+	public void sampleMissions(State s) {
+		// get a working copy of this matrix
+		double[][] player_mission_suspicion = Util.clone(this.player_mission_suspicion);
+
+		// determine how many missions we need to sample for each player
+		int[] ns = new int[players.length];
+		for (int i = 0; i < players.length; i++) {
+			PlayerState ps = s.playerState(i);
+			ns[i] = ps.missions.size();
+			ps.missions = EnumSet.noneOf(Mission.class);
+		}
+
+		// players' missions will be removed from this as we go along
+		s.missions = EnumSet.allOf(Mission.class);
+
+		// make sure our own drawn_missions aren't assigned to other players
+		Set<Mission> ms = s.playerState(player).drawn_missions;
+		if (ms != null) {
+			for (Mission m: ms) {
+				s.missions.remove(m);
+				for (int k = 0; k < players.length; k++)
+					player_mission_suspicion[m.ordinal()][k] = 0;
+			}
+		}
+		
+		sampling: while (true) {
+			// should terminate?
+			boolean samples_needed = false;
+			for (int i = 0; i < players.length; i++) {
+				if (ns[i] > 0) {
+					samples_needed = true;
+					break;
+				}
+			}
+			if (!samples_needed)
+				break sampling;
+
+			// determine non-normalized joint probability distribution
+			double total = 0;
+			double[][] jpd = Util.clone(player_mission_suspicion);
+			for (int i = 0; i < jpd.length; i++) {
+				for (int j = 0; j < jpd[i].length; j++) {
+					// more probability to the players that need more missions,
+					// less probability to those that have more railways (to avoid bias,
+					// and to generally take the average of the relevance of the claimed
+					// railways)
+					if (ns[j] == 0 && jpd[i][j] == Double.POSITIVE_INFINITY) {
+						System.err.println("inconsistent certainty");
+						throw new RuntimeException();
+					}
+					jpd[i][j] *= ns[j];
+					if (jpd[i][j] < 0 || Double.isNaN(jpd[i][j])) {
+						System.err.println("negative or NaN weight in non-normalized distribution: "+jpd[i][j]+" (originally "+player_mission_suspicion[i][j]+")");
+						System.err.println("last factor: "+ns[j]+"/"+(1 + s.playerState(j).railways.size()));
+						System.err.println("original distribution: "+Arrays.deepToString(player_mission_suspicion));
+						throw new RuntimeException();
+					}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 362:367, 473:478.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.